### DISCUSSION TITLE
Image Classification Datasets Overview - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#3382] [Docs]
# Image Classification Datasets Overview - Ultralytics YOLOv8 Docs

Learn how torchvision organizes classification image datasets. Use this code to create and train models. CLI and Python code shown.

https://docs.ultralytics.com/datasets/classify/

**Comment by impfect** [comment]
The dataset format must have a Train dir AND Test dir else you will get an error TypeError: expected str, bytes or os.PathLike object, not NoneType`.

**Reply by timlawrenz** [reply]
Some also might need a Val dir.

**Reply by blueclowd** [reply]
Same in my training, two subfolders 'train' and 'test' are necessary under the dataset folder. The data directory visualisation at the top of the page is kinda misleading

**Reply by pderrenger** [reply]
@blueclowd thank you for bringing this to our attention. The dataset directory structure indeed requires both 'train' and 'test' subfolders, and optionally a 'val' subfolder for validation. The documentation aims to provide clear guidance, and we will review the visual representation to ensure it accurately reflects the necessary directory structure for training with YOLOv8. Your feedback is valuable in helping us improve. If you have any further questions or encounter other issues, please don't hesitate to reach out. Happy training! 😊🚀

**Comment by deepanalyticszlc** [comment]
If using our own dataset, how can we include background data?

**Reply by vineetvermait** [reply]
+1

**Reply by Bleach665** [reply]
++

**Reply by pderrenger** [reply]
@Bleach665 when training object detection models like YOLOv8, it's common to include examples of "background" or "negative" images in your dataset. These are images that do not contain any of the objects of interest. Including such images helps the model learn what not to detect and reduces false positives.

To include background data in your custom dataset, you can follow these steps:

1. **Collect Background Images**: Gather images that represent the background scenes without any objects that you want the model to detect.

2. **Annotate Background Images**: For these background images, you should create annotation files similar to those for the positive images but without any bounding box annotations. Essentially, the annotation files will be empty or contain no object entries.

3. **Integrate with Dataset**: Include these background images and their corresponding annotation files in your dataset directory, following the same structure as your positive samples.

4. **Update Dataset Configuration**: Ensure your dataset configuration file (often a YAML file) includes the paths to the background images and annotations.

5. **Train the Model**: Proceed with training your model as usual. The model will learn from both the positive samples with objects and the negative samples without objects.

By including background data in your training process, the model should become more robust in distinguishing between relevant objects and irrelevant background content. 🌟

Remember to balance the number of background images with the number of positive samples to prevent the model from becoming biased towards predicting the background class too often.

**Comment by ChenJian7578** [comment]
hello， if i want to train myself dataset, what should i do about the yaml doc

**Reply by blueclowd** [reply]
In my classification training I didn't involve .yaml. Maybe its optional for classification but necessary for other tasks like detection and segmentation.

**Reply by pderrenger** [reply]
@blueclowd to train a custom dataset for image classification with YOLOv8, you'll need to organize your data in a specific format. While a `.yaml` file is not strictly necessary for classification tasks, it can be useful for defining dataset paths and parameters.

Here's a general approach to prepare your custom dataset:

1. **Organize Your Dataset**: Arrange your images into folders, with one folder for each class. The folder name should be the class name.

2. **Dataset Configuration**: Although not required, you can create a `.yaml` file to define your dataset paths and other parameters if you wish to keep your training configuration organized. Here's an example structure for the `.yaml`:

    ```yaml
    path: ../datasets/my_dataset  # dataset root dir
    train: train  # train images (relative to 'path') 256 images
    val: val  # val images (relative to 'path') 256 images
    nc: 2  # number of classes
    names: ['class0', 'class1']  # class names
    ```

3. **Training**: Use the `train` mode in YOLOv8 to start training your model. If you've created a `.yaml` file, you can specify it in the `data` argument. If not, you can directly pass the path to your dataset.

    ```python
    from ultralytics import YOLO

    # If using a .yaml file
    model = YOLO('yolov8n-cls.yaml')  # specify your model config
    results = model.train(data='my_dataset.yaml', epochs=5)

    # If not using a .yaml file, specify dataset directory directly
    model = YOLO('yolov8n-cls.yaml')  # specify your model config
    results = model.train(data='../datasets/my_dataset', epochs=5)
    ```

4. **Validation**: After training, you can validate your model using the `val` mode.

    ```python
    model.val(data='my_dataset.yaml')  # If using a .yaml file
    model.val(data='../datasets/my_dataset')  # If not using a .yaml file
    ```

Remember to replace `'yolov8n-cls.yaml'` with the appropriate model configuration file for your use case, and adjust the paths and class names to match your dataset. 🚀

For more detailed instructions and examples, please refer to the [Ultralytics Docs](https://docs.ultralytics.com).

**Comment by 13185742215** [comment]
What is the formats of these datasets? How can we costomize own datasets?

**Reply by pderrenger** [reply]
@13185742215 the datasets for image classification typically come in a simple format where images are organized into directories, with each directory name corresponding to the class label of the images contained within. For example, if you have a dataset with classes "cat" and "dog," your directory structure might look like this:

```
/dataset
    /cat
        cat1.jpg
        cat2.jpg
        ...
    /dog
        dog1.jpg
        dog2.jpg
        ...
```

To customize your own dataset for use with YOLOv8, you should organize your images in a similar structure. Once your images are organized, you can create a `.yaml` file that specifies the paths to your train and validation datasets, as well as the number of classes and class names. Here's an example of what this `.yaml` file might look like:

```yaml
# custom_dataset.yaml
path: /path/to/dataset  # dataset root dir
train: train  # train images (relative to 'path') 512 images
val: val  # val images (relative to 'path') 128 images
nc: 2  # number of classes
names: ['cat', 'dog']  # class names
```

You can then use this `.yaml` file to train your model with the `train` mode in YOLOv8. For more detailed instructions on dataset preparation and training, please refer to the [Train](https://docs.ultralytics.com/modes/train) section in our documentation. 😊🚀

**Comment by Vera370** [comment]
What is the dataset format for multi-label image classification? How to use YoloV8 for multi-label classification tasks?

**Reply by pderrenger** [reply]
@Vera370 for multi-label image classification, the dataset format typically includes an image and a corresponding list of labels that are applicable to that image. Each image can have multiple labels associated with it, unlike single-label classification where each image is assigned to one class.

To use YOLOv8 for multi-label classification tasks, you would follow these steps:

1. **Prepare your dataset**: Organize your images and create a label file for each image. The label file should contain the indices of all applicable classes for that image, possibly in a space-separated format.

2. **Define your model**: Use a YOLOv8 model configuration that supports classification. You can start with a predefined model configuration or create a custom `.yaml` file that defines the model architecture suitable for classification.

3. **Train the model**: Use the `train` mode to train your YOLOv8 model on the multi-label classification dataset. Make sure to set the `task` to `classify` and adjust the hyperparameters as needed for your specific dataset.

4. **Evaluate the model**: After training, use the `val` mode to evaluate the model's performance on a validation set. This will help you understand how well the model generalizes to unseen data.

5. **Make predictions**: Use the `predict` mode to classify new images using the trained model. The model will output the predicted labels for each image.

6. **Export the model**: If needed, you can export the trained model to various formats using the `export` mode for deployment.

Remember to set the `single_cls` flag to `False` since you are dealing with multi-label classification. Also, ensure that your dataset and model configuration are compatible with multi-label classification tasks.

For more detailed instructions and examples, please refer to the Ultralytics documentation on training and using models for classification tasks.
